tenders/hvt/hvt_cpu_aarch64.c

```diff
@@ -91,13 +91,14 @@ void aarch64_setup_memory_mapping(uint8_t *mem, uint64_t mem_size)
     for (; paddr < mem_size; paddr += PMD_SIZE, pmd++)
         *pmd = paddr | PROT_SECT_NORMAL_EXEC;
 
-    /* Link pmd tables to pud[0..3] */
-    *pud++ = AARCH64_PMD_PGT_BASE | PGT_DESC_TYPE_TABLE;
-    *pud++ = (AARCH64_PMD_PGT_BASE + 0x1000) | PGT_DESC_TYPE_TABLE;
-    *pud++ = (AARCH64_PMD_PGT_BASE + 0x2000) | PGT_DESC_TYPE_TABLE;
-    *pud++ = (AARCH64_PMD_PGT_BASE + 0x3000) | PGT_DESC_TYPE_TABLE;
+    /* Link pmd tables (PMD0, PMD1, PMD2, PMD3) to pud[0] ~ pud[3] */
+    phys_pmd_tbl = AARCH64_PMD_PGT_BASE;
+    for (paddr = 0; paddr < mem_size;
+         paddr += PUD_SIZE, pud++, phys_pmd_tbl += PAGE_SIZE)
+        *pud = phys_pmd_tbl | PGT_DESC_TYPE_TABLE;
 
     /* Mapping MMIO */
+    pud += ((AARCH64_MMIO_BASE - paddr) >> PUD_SHIFT);
```

Comment:
This is a bit dangerous, if someone changes the code later in a way that can cause `(AARCH64_MMIO_BASE - paddr)` to go negative, the result will be bogus. Perhaps add an `assert(paddr <= AARCH64_MMIO_BASE);` just before?